TombEngine/Objects/TR3/Entity/tr3_wasp.cpp

```diff
@@ -0,0 +1,286 @@
+	}
+
+	static void DoWaspEffects(short itemNumber, ItemInfo& item)
+	{
+		constexpr auto PARTICLE_EFFECT_COUNT = 2;
+
+		// Spawn light.
+		auto pos = GetJointPosition(&item, WaspVenomSackBite.meshNum, WaspVenomSackBite.Position);
+		TriggerDynamicLight(
+			pos.x, pos.y, pos.z, 10, 
+			WaspVenomSackLightColor.x * 255,
+			WaspVenomSackLightColor.y * 255,
+			WaspVenomSackLightColor.z * 255);
+
+		// Spawn wasp effect twice.
+		for (int i = 0; i < PARTICLE_EFFECT_COUNT; i++)
+			SpawnWaspParticle(itemNumber);
+	}
+
+	void InitialiseWaspMutant(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		InitialiseCreature(itemNumber);
+		SetAnimation(&item, WASP_STATE_IDLE); // Start as flying idle.
+	}
+
+	// NOTE: AI_MODIFY not allow the wasp to land,
+	// and if he start at the land state (which is by default the case),
+	// he will be forced to fly !
+	void WaspMutantControl(short itemNumber)
+	{
+		if (!CreatureActive(itemNumber))
+			return;
+
+		auto& item = g_Level.Items[itemNumber];
+		auto& creature = *GetCreatureInfo(&item);
+
+		short headingAngle = 0;
+
+		if (item.HitPoints <= 0)
+		{
+			switch (item.Animation.ActiveState)
+			{
+			case WASP_STATE_FALL:
+				if (item.Pose.Position.y >= item.Floor)
+				{
+					item.Pose.Position.y = item.Floor;
+					item.Animation.IsAirborne = false;
+					item.Animation.Velocity.y = 0.0f;
+					item.Animation.TargetState = WASP_STATE_DEATH;
+				}
+
+				break;
+
+			case WASP_STATE_DEATH:
+				item.Pose.Position.y = item.Floor;
+				break;
+
+			default:
+				SetAnimation(&item, WASP_STATE_FALL);
+				item.Animation.IsAirborne = true;
+				item.Animation.Velocity = Vector3::Zero;
+				break;
+			}
+
+			item.Pose.Orientation.x = 0;
+		}
+		else
+		{
+			if (item.AIBits)
+				GetAITarget(&creature);
+
+			AI_INFO ai;
+			CreatureAIInfo(&item, &ai);
+
+			GetCreatureMood(&item, &ai, true);
+			CreatureMood(&item, &ai, true);
+
+			headingAngle = CreatureTurn(&item, creature.MaxTurn);
+
+			switch (item.Animation.ActiveState)
+			{
+			case WASP_STATE_IDLE:
+				creature.MaxTurn = WASP_LAND_TURN_RATE_MAX;
+				item.Pose.Position.y = item.Floor;
+
+				if (item.HitStatus ||
+					ai.distance < WASP_TAKE_OFF_RANGE ||
+					creature.HurtByLara ||
+					item.AIBits == MODIFY)
+				{
+					item.Animation.TargetState = WASP_STATE_IDLE_TO_FLY_IDLE;
+				}
+
+				break;
+
+			case WASP_STATE_FLY_IDLE_TO_IDLE:
+				creature.MaxTurn = WASP_LAND_TURN_RATE_MAX;
+
+				item.Pose.Position.y += WASP_IDLE_TO_FLY_VELOCITY;
+				if (item.Pose.Position.y >= item.Floor)
+					item.Pose.Position.y = item.Floor;
+
+				break;
+
+			case WASP_STATE_FLY_IDLE:
+				creature.MaxTurn = WASP_AIR_TURN_RATE_MAX;
+				creature.Flags = 0;
+
+				if (item.Animation.RequiredState)
+				{
+					item.Animation.TargetState = item.Animation.RequiredState;
+				}
+				// NOTE: This causes the wasp to wait until probability is valid or
+				// the player has hit the wasp to move forward, which is conceptually bad.
```

Comment:
Ehh yeah it is conceptually bad imo. I couldn't get t to work properly. The spawner stuff can be perfectly replicated in Lua using the SetOnKill() (or whatev it was called) and Moveable.new() stuff.

Reply:
Will change it :)

Reply:
Done